From 85062d02b4759b3b3e027e82dedc87fdf84d6c69 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sat, 31 Oct 2015 21:39:35 +0000
Subject: [PATCH 1/4] MinGW-w64: Two fixes for 'unusual' files.

Handle isatty() files better.
Pretend 'nul' is '/dev/null' which is necessary for libtool.
---
 binutils/bucomm.c  | 19 +++++++++++++++++--
 binutils/elfedit.c | 15 ++++++++++++++-
 binutils/readelf.c | 15 ++++++++++++++-
 3 files changed, 45 insertions(+), 4 deletions(-)

diff --git a/binutils/bucomm.c b/binutils/bucomm.c
index 4621e60..e6a20eb 100644
--- a/binutils/bucomm.c
+++ b/binutils/bucomm.c
@@ -574,6 +574,14 @@ off_t
 get_file_size (const char * file_name)
 {
   struct stat statbuf;
+  int f, t;
+  t = -1;
+  f = open (file_name, O_RDONLY | O_BINARY);
+  if (f != 0)
+    {
+      t = isatty (f);
+      close (f);
+    }
 
   if (stat (file_name, &statbuf) < 0)
     {
@@ -583,8 +591,15 @@ get_file_size (const char * file_name)
 	non_fatal (_("Warning: could not locate '%s'.  reason: %s"),
 		   file_name, strerror (errno));
     }
-  else if (! S_ISREG (statbuf.st_mode))
-    non_fatal (_("Warning: '%s' is not an ordinary file"), file_name);
+  else if (! S_ISREG (statbuf.st_mode) || t > 0)
+    {
+#ifdef _WIN32
+      /* libtool passes /dev/null and checks for /dev/null in the output */
+      if (stricmp (file_name, "nul") == 0)
+        file_name = "/dev/null";
+#endif
+      non_fatal (_("Warning: '%s' is not an ordinary file"), file_name);
+    }
   else if (statbuf.st_size < 0)
     non_fatal (_("Warning: '%s' has negative size, probably it is too large"),
                file_name);
diff --git a/binutils/elfedit.c b/binutils/elfedit.c
index 02b9c98..fb0eadf 100644
--- a/binutils/elfedit.c
+++ b/binutils/elfedit.c
@@ -483,6 +483,14 @@ static int
 check_file (const char *file_name, struct stat *statbuf_p)
 {
   struct stat statbuf;
+  int f, t;
+  t = -1;
+  f = open (file_name, O_RDONLY | O_BINARY);
+  if (f != 0)
+    {
+      t = isatty (f);
+      close (f);
+    }
 
   if (statbuf_p == NULL)
     statbuf_p = &statbuf;
@@ -497,8 +505,13 @@ check_file (const char *file_name, struct stat *statbuf_p)
       return 1;
     }
 
-  if (! S_ISREG (statbuf_p->st_mode))
+  if (! S_ISREG (statbuf_p->st_mode) || t > 0)
     {
+#ifdef _WIN32
+      /* libtool passes /dev/null and checks for /dev/null in the output */
+      if (stricmp (file_name, "nul") == 0)
+        file_name = "/dev/null";
+#endif
       error (_("'%s' is not an ordinary file\n"), file_name);
       return 1;
     }
diff --git a/binutils/readelf.c b/binutils/readelf.c
index 70a84e5..229a5f5 100644
--- a/binutils/readelf.c
+++ b/binutils/readelf.c
@@ -16523,6 +16523,14 @@ process_file (char * file_name)
   struct stat statbuf;
   char armag[SARMAG];
   int ret;
+  int f, t;
+  t = -1;
+  f = open (file_name, O_RDONLY | O_BINARY);
+  if (f != 0)
+    {
+      t = isatty (f);
+      close (f);
+    }
 
   if (stat (file_name, &statbuf) < 0)
     {
@@ -16534,8 +16542,13 @@ process_file (char * file_name)
       return 1;
     }
 
-  if (! S_ISREG (statbuf.st_mode))
+  if (! S_ISREG (statbuf.st_mode) || t > 0)
     {
+#ifdef _WIN32
+      /* libtool passes /dev/null and checks for /dev/null in the output */
+      if (stricmp (file_name, "nul") == 0)
+        file_name = "/dev/null";
+#endif
       error (_("'%s' is not an ordinary file\n"), file_name);
       return 1;
     }
-- 
2.6.2

